chalk-integration/src/lowering.rs

```diff
@@ -96,110 +97,80 @@ enum ApplyTypeLookup {
-                    .intern(interner)
-                    .cast(interner)),
-                chalk_ir::VariableKind::Lifetime => Ok(chalk_ir::LifetimeData::BoundVar(*b)
-                    .intern(interner)
-                    .cast(interner)),
-                chalk_ir::VariableKind::Const(ty) => {
-                    Ok(b.to_const(interner, ty.clone()).cast(interner))
-                }
-            };
-        }
-
-        if let Some(id) = self.adt_ids.get(&name.str) {
-            let k = self.adt_kind(*id);
+        let apply = |k: &TypeKind, type_name: chalk_ir::TypeName<ChalkIr>| {
             if k.binders.len(interner) > 0 {
-                return Err(RustIrError::IncorrectNumberOfTypeParameters {
+                Err(RustIrError::IncorrectNumberOfTypeParameters {
                     identifier: name.clone(),
                     expected: k.binders.len(interner),
                     actual: 0,
-                });
-            } else {
-                return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
-                    name: chalk_ir::TypeName::Adt(*id),
-                    substitution: chalk_ir::Substitution::empty(interner),
                 })
-                .intern(interner)
-                .cast(interner));
-            };
-        }
-
-        if let Some(id) = self.fn_def_ids.get(&name.str) {
-            let k = self.fn_def_kind(*id);
-            if k.binders.len(interner) > 0 {
-                return Err(RustIrError::IncorrectNumberOfTypeParameters {
-                    identifier: name.clone(),
-                    expected: k.binders.len(interner),
-                    actual: 0,
-                });
             } else {
-                return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
-                    name: chalk_ir::TypeName::FnDef(*id),
+                Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
+                    name: type_name,
                     substitution: chalk_ir::Substitution::empty(interner),
                 })
                 .intern(interner)
-                .cast(interner));
+                .cast(interner))
             }
-        }
+        };
 
-        if let Some(id) = self.closure_ids.get(&name.str) {
-            let k = self.closure_kind(*id);
-            if k.binders.len(interner) > 0 {
-                return Err(RustIrError::IncorrectNumberOfTypeParameters {
-                    identifier: name.clone(),
-                    expected: k.binders.len(interner),
-                    actual: 0,
-                });
-            } else {
-                return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
-                    name: chalk_ir::TypeName::Closure(*id),
-                    // See note in `program`. Unlike rustc, we store upvars separately.
-                    substitution: chalk_ir::Substitution::empty(interner),
-                })
+        if let Ok(lookup) = self.lookup_apply_type(name) {
+            match lookup {
```

Comment:
Maybe use a single match for this? Then the arms be `Ok(ApplyTypeLookup::Param(p))`, `Ok(ApplyTypeLookup::Adt(id))`,..., `Err(_)`